### PR TITLE
Witgen: Handle Block-to-Block calls

### PR DESCRIPTION
**Pull request by georgwiese**
With this PR, witness generation is able to generate witnesses for block machines which call other block machines.

The required changes are:
- Some changes in machine detection to actually detect the machines properly in this case
- `IdentityProcessor` now holds a `Vec<&mut KnownMachine<'a, T>>` instead of `machines: Vec<KnownMachine<'a, T>>`, wrapped in a new type.
- When a sub-machine is called, the current machine is removed from the vector of machines and the other machines are passes to the callee. This way, a machine can call other machines, but there can't be any cycles. With #515 implemented, cycles can be supported and the code can be simplified.

Note that this creates a new (small) vector of mutable references of machines for each row (this can change once `Generator` computes an entire block instead of single rows) and once per lookup and machine.

That's no performance issue though, according to the benchmark:
```
keccak-executor-benchmark/keccak
                        time:   [31.351 s 31.752 s 32.184 s]
                        change: [-4.5602% -2.1870% +0.2718%] (p = 0.11 > 0.05)
```

### CODE REVIEW

File: executor/src/witgen/identity_processor.rs [hunk 1]
```diff
@@ -111,15 +153,18 @@ impl<'a, 'b, T: FieldElement> IdentityProcessor<'a, 'b, T> {
...
-            if let Some(result) = m.process_plookup(
-                self.fixed_data,
-                self.fixed_lookup,
-                identity.kind,
-                &left,
-                &identity.right,
-            ) {
+        for i in 0..self.machines.len() {
+            if let Some(result) = self.machines.split(i, |current, others| {
+                // TODO also consider the reasons above.
```
Comment (georgwiese) [inline executor/src/witgen/identity_processor.rs:158]: @chriseth Do you actually know what this comment means? I think it's been copied around quite a bit recently, but I'm not sure if it's still relevant.

Comment (chriseth) [reply executor/src/witgen/identity_processor.rs:166]: At some point, we were ignoring failures from some identities. I'm not sure if we are still collecting reasons for failure. We were just ignoring some reasons and did not put them into the vector where we collected the failures.

File: executor/src/witgen/identity_processor.rs [hunk 2]
```diff
@@ -5,25 +5,75 @@ use itertools::{Either, Itertools};
...
 };
 
+/// A list of mutable references to machines.
+pub struct Machines<'a, 'b, T: FieldElement> {
+    machines: Vec<&'b mut KnownMachine<'a, T>>,
+}
+
+impl<'a, 'b, T: FieldElement> Machines<'a, 'b, T> {
+    /// Applies a given function to the ith machine and all other machines.
+    /// With #515 implemented, this could be removed as there would not be a need to split the
```
Comment (chriseth) [inline executor/src/witgen/identity_processor.rs:24]: Can you summarize 515? I don't really like code referencing github issues because they might not be available.

File: executor/src/witgen/identity_processor.rs [hunk 3]
```diff
@@ -5,25 +5,75 @@ use itertools::{Either, Itertools};
...
+    /// list of machines.
+    pub fn split<'c, R>(
+        &'c mut self,
+        i: usize,
```
Comment (chriseth) [inline executor/src/witgen/identity_processor.rs:28]:
```suggestion
        index: usize,
```

File: executor/src/witgen/identity_processor.rs [hunk 4]
```diff
@@ -5,25 +5,75 @@ use itertools::{Either, Itertools};
...
     EvalResult, EvalValue, FixedData, IncompleteCause,
 };
 
+/// A list of mutable references to machines.
+pub struct Machines<'a, 'b, T: FieldElement> {
+    machines: Vec<&'b mut KnownMachine<'a, T>>,
+}
+
+impl<'a, 'b, T: FieldElement> Machines<'a, 'b, T> {
+    /// Applies a given function to the ith machine and all other machines.
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/identity_processor.rs:23]: It will probably make the borrow checking much more complicated, will it?

File: executor/src/witgen/identity_processor.rs [hunk 5]
```diff
@@ -5,25 +5,75 @@ use itertools::{Either, Itertools};
...
+}
+
+impl<'a, 'b, T: FieldElement> From<Vec<&'b mut KnownMachine<'a, T>>> for Machines<'a, 'b, T> {
+    fn from(machines: Vec<&'b mut KnownMachine<'a, T>>) -> Self {
+        Self { machines }
+    }
+}
+
+impl<'a, 'b, T: FieldElement> From<&'b mut Vec<KnownMachine<'a, T>>> for Machines<'a, 'b, T> {
+    fn from(machines: &'b mut Vec<KnownMachine<'a, T>>) -> Self {
```
Comment (chriseth) [inline executor/src/witgen/identity_processor.rs:58]: Is it possible to generalize this from a vector to arbitrary collections? Not a blocker though.

File: executor/src/witgen/machines/machine_extractor.rs [hunk 6]
```diff
@@ -57,25 +57,26 @@ pub fn split_out_machines<'a, T: FieldElement>(
             base_identities.iter().cloned().partition(|i| {
                 // The identity has at least one machine witness, but
```
Comment (chriseth) [inline executor/src/witgen/machines/machine_extractor.rs:58]: Can you update this documentation? It does not fit any more to the fact that we only consider the 'left' part.
The idea is that lookups from that machine are part of the machine, but not those "to" the machine, right?

Comment (chriseth) [reply executor/src/witgen/machines/machine_extractor.rs:58]: At the same time, the implementation of `all_row_connected_witnesses` should consider all witnesses on the left hand side of a lookup to be row-connected but also all witnesses on the right hand side of a lookup (but they sides are not row-connected), right?

File: executor/src/witgen/machines/machine_extractor.rs [hunk 7]
```diff
@@ -57,25 +57,26 @@ pub fn split_out_machines<'a, T: FieldElement>(
...
             .cloned()
             .filter(|i| {
-                refs_in_identity(i)
+                refs_in_selected_expressions(&i.right)
                     .intersection(&machine_witnesses)
                     .next()
                     .is_some()
             })
             .collect::<Vec<_>>();
+        assert!(connecting_identities.contains(id));
```
Comment (chriseth) [inline executor/src/witgen/machines/machine_extractor.rs:76]: What if `id` is not a lookup?

File: executor/src/witgen/machines/machine_extractor.rs [hunk 8]
```diff
@@ -57,25 +57,26 @@ pub fn split_out_machines<'a, T: FieldElement>(
...
                 // all referenced witnesses are machine witnesses.
-                let all_refs = &refs_in_identity(i) & (&all_witnesses);
+                let all_refs = &refs_in_selected_expressions(&i.left) & (&all_witnesses);
                 !all_refs.is_empty() && all_refs.is_subset(&machine_witnesses)
             });
         base_identities = remaining_identities;
         remaining_witnesses = &remaining_witnesses - &machine_witnesses;
 
-        let connecting_identities = base_identities
+        let connecting_identities = identities
```
Comment (chriseth) [inline executor/src/witgen/machines/machine_extractor.rs:66]: Maybe add a comment: These are the "calling" identities.

Comment (chriseth) [reply executor/src/witgen/machines/machine_extractor.rs:66]: Actually a bit unrelated to this PR: So we are actually determining all the "calling" identities, but IIRC, we do not automatically query the machine related to the connecting identity, do we, instead we iterate through all machines - or is that fixed already?

Comment (georgwiese) [reply executor/src/witgen/machines/machine_extractor.rs:71]: Yes, that is still the case!

File: executor/src/witgen/machines/machine_extractor.rs (hunk 7 above)
Comment (georgwiese) [reply executor/src/witgen/machines/machine_extractor.rs:81]: Then the loop exits early:
```rust
        if lookup_witnesses.is_empty() {
            continue;
        }
```

File: executor/src/witgen/machines/machine_extractor.rs (hunk 6 above)
Comment (georgwiese) [reply executor/src/witgen/machines/machine_extractor.rs]: Sure, I updated the comment and elaborated a bit. Also, yes, `all_row_connected_witnesses` should work exactly as you describe.

**Review by georgwiese — COMMENTED** [review]
Cool, thanks for the review!

**Review by chriseth — APPROVED** [review]
